### PR TITLE
remove legacy perm check

### PR DESCRIPTION
**Pull request by snopoke**
While looking at the auth code in the user model I noticed a number of inconsistencies and unnecessary complexities. This PR is the first step in bringing about more consistency to the following user methods:

```python
user.get_membership
user.is_member_of
user.is_domain_admin
user.has_permission
user.get_role
```

Best reviewed by commit 🐡 

## Technical Summary
The primary commit in this PR a420f4f80e86790352e65902a800538b0c9e2710 which removes the `restrict_global_admin` param from `user.has_permission`. The usage of this param was removed in b5afa53b1052834cb4a8fe69a059301fd5d3692c.

The commits before that are adding tests and small changes to make the tests pass. Each change is in it's own commit including any changes to the test cases which indicate any changes the function behaviour.

The commits after a420f4f80e86790352e65902a800538b0c9e2710 are minor refactors.

The tests have some TODO comments where the functionality appears to be inconsistent between different functions. I'd like to fix those but I will do so in a separate PR so to avoid introducing too much change at once. I'd also like these tests to get into the codebase for the sake of future refactors.

### Safety story
The code appears to have reasonable test coverage already: https://app.codecov.io/gh/dimagi/commcare-hq/blob/master/corehq/apps/users/models.py

I have also added a number of tests that document the current functionality and test more edge cases.

There are a few small changes to functionality as shown below but I don't think these are edge cases that are relevant. The functionality changes are as a result of more consistent handling of null domain arguments.

***is_member_of* called for global admin user with null domain:**
```diff
- global_admin.is_member_of(None) == True
+ global_admin.is_member_of(None) == False
```

***has_permissoin* called for global admin user with null domain**
```diff
- global_admin.has_permission(None, 'report_an_issue') == True
+ global_admin.has_permission(None, 'report_an_issue') == False
```

### Automated test coverage
Added

### QA Plan
I don't plan on doing any formal QA of this.

### Migrations
NA

### Rollback instructions

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

### CODE REVIEW

File: corehq/apps/users/models.py [hunk 1]
```diff
@@ -562,13 +562,10 @@ def transfer_domain_membership(self, domain, to_user, create_record=False, is_ad
...
     def is_domain_admin(self, domain=None):
+        # this is a hack needed because we can't pass parameters from views
+        domain = domain or getattr(self, 'current_domain', None)
         if not domain:
-            # hack for template
-            if hasattr(self, 'current_domain'):
-                # this is a hack needed because we can't pass parameters from views
-                domain = self.current_domain
-            else:
-                return False  # no domain, no admin
```
Comment (snopoke) [inline corehq/apps/users/models.py:568]: this block could still end up with a null domain if `self.current_domain = None`

**Comment by snopoke** [comment]
Potentially related build failures

File: corehq/apps/users/models.py (hunk 1 above)
Comment (esoergel) [reply corehq/apps/users/models.py:568]: Writing this out because I'm having difficulty following.  When this is called without a `domain` arg there are two big forks - whether or not this was called inside a normal request, where `current_domain` would be set, possibly to `None` if it's not a domain page.

For a global admin, the previous behavior was:
If `current_domain` is set, always allow, as long as this isn't known to be a domain which restricts superusers.  If it's not set, always block.

The new behavior is:
If `current_domain` is set, allow as long as the domain is known and doesn't restrict superusers.  Otherwise block.

For a non-global admin, I don't think there's a change (assuming `get_domain_membership(None)` returns `None`)

That sounds like a fairly safe change - I think the failure mode would be that superusers could in theory be denied permission if accessing something that can't find the domain, whereas they couldn't before.

**Review by esoergel — APPROVED** [review]
This auth code is so hard to follow, thanks for cleaning it up a bit.  I'm a little nervous in general in making changes to this sort of thing, but I didn't see anything that looked off.

File: corehq/apps/users/models.py [hunk 2]
```diff
@@ -438,6 +438,9 @@ class Meta(object):
 class IsMemberOfMixin(DocumentSchema):
 
     def _is_member_of(self, domain, allow_enterprise):
+        if not domain:
+            return False
```
[the remark this answers is not in the document]
Comment (snopoke) [reply corehq/apps/users/models.py:442]: That would also be my preference but it's hard to tell if there are places that might call it with a null domain. What I could do is add a soft assert (in a future PR) to try and determine that.

There is also mixed functionality between the different methods and how they handle it.

File: corehq/apps/users/tests/test_authorization.py [hunk 3]
```diff
@@ -251,10 +251,6 @@ def test_is_domain_admin__non_admin_role__domain_restricts_superusers(self, _moc
     def test_is_member_of__no_membership(self):
         self.assertTrue(self.user.is_member_of('other'))
 
-    def test_is_member_of__null_domain(self):
-        # TODO: shouldn't this be False?
-        self.assertTrue(self.user.is_member_of(None))
```
[the remark this answers is not in the document]
Comment (snopoke) [reply corehq/apps/users/tests/test_authorization.py]: This is for a superuser and it checked the superuser status before checking if the domain was valid.

File: corehq/apps/users/models.py (hunk 1 above)
Comment (snopoke) [reply corehq/apps/users/models.py:568]: That's correct. This change handles an empty domain the same as None (return False)

> whether or not this was called inside a normal request, where current_domain would be set, possibly to None if it's not a domain page.

`user.current_domain` should never be set to None: https://github.com/dimagi/commcare-hq/search?q=%22user.current_domain+%3D%22&type=

**Comment by snopoke** [comment]
> I'm a little nervous in general in making changes to this sort of thing, but I didn't see anything that looked off.

Likewise which is why I added the test suite before making any changes. If you think I've missed any scenarios in the test suite I can add them in.

**Comment by snopoke** [comment]
I'm going to put this on staging for a bit just to rule out any major issues.

**Comment by snopoke** [comment]
This has been on staging since March 7 without issue so I'm comfortable merging.
